[✨ feat] 전역 예외 처리 기능 구현

jsoonworld opened the pull request:
# 📄 Work Description  
- 공통 API 응답 포맷을 위한 `ApiResponse<T>` 클래스 구현  
- 컨트롤러 전역 예외 처리를 담당하는 `ExceptionHandler` 클래스 구현  
- JSON 파싱 오류, Bean Validation 실패, 비즈니스/도메인 예외 등 다양한 상황에 대해 일관된 형식으로 에러 응답을 반환하도록 구성  
- 클라이언트가 예측 가능한 구조의 응답을 받을 수 있도록 표준화된 응답 형태 제공  
- `com.terning.server.kotlin.ui.api` 패키지 하위에 위치시켜, API UI 계층의 책임을 명확히 분리함

# 💭 Thoughts  
- 다양한 예외 상황에서 일관된 구조(`ApiResponse`)로 응답을 내려주는 방식은 클라이언트 입장에서 에러 핸들링을 훨씬 수월하게 만들어줄 수 있을 것 같아 도입하게 되었습니다.  
- 특히 JSON 파싱 오류(`HttpMessageNotReadableException`)는 사용자 입력에서 자주 발생하기 때문에, `MismatchedInputException`과 `InvalidFormatException`을 구분하여 조금 더 구체적이고 명확한 메시지를 전달하려고 신경 썼습니다.  
- 다만 `exception.path.lastOrNull()?.fieldName`을 통해 필드명을 추출하는 방식이 모든 케이스에서 안정적으로 동작할지 확신이 서지 않아, 필드명이 누락될 수 있는 경우에 더 나은 처리 방식이 있다면 리뷰어님의 의견을 듣고 싶습니다 🙏  
- `@ExceptionHandler`에 정의되는 예외가 점점 많아지면 클래스가 복잡해질 수 있을 것 같아, 예외별로 분리하거나 다른 방식으로 관리하는 것이 좋을지 고민이 되었습니다. 리뷰어님께서는 어떤 기준으로 유지하거나 분리하는 게 좋다고 생각하시는지 궁금합니다!  
- 현재는 모든 예외 메시지를 문자열 하나로 응답하고 있지만, 여러 필드에 대한 검증 실패 메시지를 리스트나 객체 형태로 제공하는 것이 더 효과적일지도 고민 중입니다. 이 부분에 대해서도 의견을 들어보고 싶어요.  
- 마지막으로 `ApiResponse`의 구조는 추후 상태 코드나 타임스탬프 등의 확장이 가능하도록 설계했는데, 혹시 개선하거나 보완할 점이 있다면 편하게 말씀해 주세요!

# ✅ Testing Result  
- `ExceptionHandler`는 Spring의 내부 예외 처리 흐름에 따라 동작하기 때문에, 단위 테스트보다는 **MockMvc를 활용한 통합 테스트**로 예외 상황을 유도하여 간접적으로 검증하는 방식이 더 적절하다고 판단했습니다.  
- 이번 PR에서는 우선 예외 처리기 자체의 구현에 집중했고, 실제 컨트롤러에서 발생하는 예외에 대해 핸들러가 정상 동작하는지 확인하는 테스트는 별도 PR로 분리하여 작성할 예정입니다.  
- `ApiResponse`는 데이터 래핑을 위한 단순 구조이며 내부 로직이 없기 때문에 별도의 테스트는 작성하지 않았습니다. 구조 변경이 없는 이상 테스트의 필요성이 높지 않다고 판단했습니다.

# 📂 Related Issue  
- closed #1

jsoonworld committed:
fix(package) : 패키지 명 오타 수정

터닝 서버 패키지 경로에서 'turning'을 'terning'으로 수정했습니다.

jsoonworld committed:
fix(test) : 테스트 클래스 패키지 경로 오타 수정

'com.turning.server.kotlin' → 'com.terning.server.kotlin'으로 변경하여 패키지명을 통일했습니다.

jsoonworld committed:
feat(ApiResponse) : 공통 응답 포맷을 위한 ApiResponse 클래스 구현

jsoonworld committed:
feat(ExceptionHandler) : 전역 예외 처리 클래스 구현

jsoonworld commented on src/main/kotlin/com/terning/server/kotlin/ui/api/ExceptionHandler.kt:
```diff
@@ -0,0 +1,70 @@
+        logger.error("message", ex)
+        val message = when (val exception = ex.cause) {
+            is MismatchedInputException -> "${exception.path.lastOrNull()?.fieldName.orEmpty()}: 널이어서는 안됩니다"
+            is InvalidFormatException -> "${exception.path.lastOrNull()?.fieldName.orEmpty()}: 올바른 형식이어야 합니다"
```
Thank you for the review! 🙌
You're absolutely right — using `exception.path.lastOrNull()?.fieldName.orEmpty()` may lead to unclear messages when the field name is missing, resulting in outputs like `": must not be null"` which could be confusing to the client.

I think your suggestion to explicitly provide a default value like `"UnknownField"` makes the error message clearer and more informative. I'll go ahead and update the message construction like this:

```kotlin
"${exception.path.lastOrNull()?.fieldName ?: "UnknownField"}: 널이어서는 안됩니다"
"${exception.path.lastOrNull()?.fieldName ?: "UnknownField"}: 올바른 형식이어야 합니다"
```

This should help improve readability and make debugging easier on the client side as well.
Thanks again for pointing that out! 😊 Let me know if there are any other areas I can refine.

jsoonworld committed:
fix(ExceptionHandler) : 필드명이 누락된 경우 기본값으로 'UnknownField' 사용하도록 수정

jsoonworld committed:
chore(gradle) : develop 브랜치의 Gradle Wrapper 파일 병합

jsoonworld committed:
chore(.gitignore) : GraalVM 및 Gradle 임시 파일 무시 설정 추가

jsoonworld committed:
test(build.gradle.kts) : 테스트 환경을 위한 H2 DB 의존성 추가

jsoonworld committed:
test(application-test.yml) : 테스트 환경 설정을 위한 H2 DB 설정 추가

jsoonworld committed:
test(ServerKotlinApplicationTests) : 테스트 환경에서 application-test.yml을 사용하도록 설정

leeeyubin commented on src/main/kotlin/com/terning/server/kotlin/ui/api/ApiResponse.kt:
```diff
@@ -0,0 +1,12 @@
+
+        fun <T> success(body: T?): ApiResponse<T> = ApiResponse(body = body)
+    }
+}
```
제가 정확히 이해를 했는지 모르겠어서 질문 몇 개 남겨보도록 하겠습니다!

1. 터닝은 아래와 같은 형식으로 응답을 주고 받는데요..! 이거와는 상관 없이 성공하면 success 함수의 body에 값을 다 넣어주는 건가요?
```
{
  "status": 200,
  "message": "메세지",
  "result": ...
}
```

leeeyubin replied:
2. 혹시 실패했을 때와 성공했을 때 전부 ApiResponse를 통해 받는 걸까요? 
그렇다면 실패 상황에서도 `body`가 nullable로 남아있어야 하다 보니 구조적으로 분리된 응답을 사용하는 건 어떤지 의견 남겨봅니다!

jsoonworld replied:
꼼꼼한 리뷰 감사합니다! 🙌

1. 처음 설계를 할 때는 `status` 값을 `HTTP` 응답 코드에서만 내려주고, 바디에는 별도로 포함시키지 않아도 된다고 생각했습니다. 실제로 `Spring`의 `ResponseEntity` 클래스도 `HTTP status` 코드는 헤더에 포함시키는 것을 기본으로 하고 있어, 중복 표현을 피하는 방향이 자연스러워 보였습니다. 하지만 유빈님 말씀을 듣고 다시 명세를 살펴보니, 바디에도 `status`, `message`, `result`를 명시적으로 포함하는 구조를 따르고 있었고, 이로 인해 응답 구조가 더 예측 가능하다는 장점이 있다는 점에서 다시 한번 방향을 재정립하게 되었습니다. 결국, `Spring`의 기본 설계 방식과는 다르지만, 명세 일관성과 클라이언트 측의 명확한 처리 기준을 위해 바디에도 `status`를 포함시키는 구조로 구현을 유지하기로 수정하였습니다!

2. 실패 응답에서는 원래 명세상 `result` 필드가 포함되지 않는 구조인데, 기존 `ApiResponse`는 `status`, `message`, `result` 세 필드가 모두 고정되어 있어 실패 시에도 "result": null이 항상 포함되는 문제가 있었습니다. 이를 개선하기 위해 `Jackson`의 `@JsonInclude(JsonInclude.Include.NON_NULL)` 어노테이션을 활용하여, `result`가 `null`인 경우에는 아예 응답 바디에서 해당 필드가 제외되도록 설정했습니다. 덕분에 명세와 더 일치하는 형태로 응답을 줄 수 있게 되었고, 클라이언트 입장에서도 보다 깔끔하고 일관된 구조로 응답을 받을 수 있게 개선되었습니다!

3. 기존 서버 코드에서는 성공과 실패 응답을 분리된 클래스로 관리하고 있었는데, 이 방식은 구조가 반복되고 코드량이 늘어나는 단점이 있었습니다. 그래서 이번에는 `ApiResponse<T>` 하나로 통합해서 성공/실패를 모두 대응하도록 정리했고, 각 상황에서 명확한 `HTTP` 상태 코드와 메시지를 포함해 처리할 수 있도록 개선했습니다.

구조가 명확해지고 코드도 줄어서 유지보수 측면에서도 더 나아진 것 같아요. 😄
혹시 이 구조에 대해 더 궁금하신 부분 있으시면 언제든지 이야기 나눠요!

leeeyubin replied:
수정된 점 확인했습니다! 자세한 설명 감사해요:)
`@JsonInclude(JsonInclude.Include.NON_NULL)` 어노테이션이 있는 지 몰랐네요!! 이 어노테이션을 활용한다면 `ApiResponse<T>` 하나로 통합해도 좋을 것 같아요!

jsoonworld committed:
chore(.gitignore) : Gradle 캐시 디렉토리 및 jdks 무시 설정 추가

leeeyubin commented on src/main/kotlin/com/terning/server/kotlin/ui/api/ExceptionHandler.kt:
```diff
@@ -0,0 +1,70 @@
+
+    private fun MethodArgumentNotValidException.messages(): String {
+        return bindingResult.fieldErrors.joinToString(", ") { "${it.field}: ${it.defaultMessage.orEmpty()}" }
+    }
```
```suggestion
    private fun MethodArgumentNotValidException.messages(): String =
        bindingResult.fieldErrors.joinToString(", ") { "${it.field}: ${it.defaultMessage.orEmpty()}" }
```
저는 return을 표현식으로 해야만 하는 병이 있어요

jsoonworld replied:
귀여운 병이네요 ㅎㅎ 저도 전염될까 봐 걱정이네요. 마스크라도 써야 할까 봐요 😷ㅎㅎ

유빈님께서 예전에 코틀린 피드백 주셨을 때, `return`보다는 표현식으로 작성하는 걸 권장해주셨던 게 인상 깊었어요. 그 이후로 코드를 작성할 때마다 표현식으로 바꿔보려는 시도를 하고 있는데요!

이번 경우엔 오히려 `return`을 명시적으로 사용하는 편이 더 낫지 않을까 하는 생각이 들었습니다.
첫 번째 이유는 가시성입니다. 표현식 형태로 작성하면 한 줄에 모든 로직이 담기다 보니, 라인이 가로로 길어지면서 오히려 읽기 어려워지는 경우가 있는 것 같아요. `return`으로 명확히 구분되면 구조적으로 더 눈에 잘 들어오고요.

두 번째는 유지보수성입니다. 결국 우리가 작성한 코드는 여러 사람이 함께 보고, 고치고, 확장해나가는 코드이기 때문에, 한눈에 구조가 파악되는 쪽이 장기적으로 봤을 때 유지보수에 더 유리하다고 느꼈어요.

물론 성능상 차이가 있고, 표현식 방식이 더 일관성 있게 보이거나 팀 스타일과 잘 맞는다면 적극적으로 도입하고 싶습니다. 다만 아직은 실용적인 면에서 `return`이 더 낫지 않나 싶어 이렇게 작성했는데, 혹시 이 부분에 대해 유빈님은 어떻게 생각하시는지 궁금해요! 😊

leeeyubin replied:
제가 생각할 때 코틀린의 장점은 간결하다는 점인 것 같아요! 표현식을 사용할 수 있다는 점이 그 이유 중 하나이기도 하구요
그래서 표현식으로 나타낼 수 있는 경우는 적극적으로 사용해주는 게 '코틀린스러운' 방식이라고 생각했습니다!

성능상 차이는 미미하겠지만 그 언어의 장점을 살려서 작성하는 습관을 기르는 게 좋을 것 같아서 의견을 제시했던 겁니당..!!

장순님께서 `return`이 더 낫다고 생각하면 사실 큰 차이는 없기 때문에 일단은 이렇게 가도 괜찮을 것 같아요~!

jsoonworld replied:
유빈님 말씀 듣고 한참 고민해봤는데, 말씀하신 표현식 스타일이 단순히 한 줄로 길게 쓰는 게 아니라, `=`을 쓰되 줄바꿈을 통해 충분히 읽기 좋게 작성할 수 있다는 의미였다는 걸 이제야 정확히 이해했어요! 😊

예를 들면 아래처럼요:

```kotlin
private fun MethodArgumentNotValidException.messages(): String =
    bindingResult.fieldErrors.joinToString(", ") {
        "${it.field}: ${it.defaultMessage.orEmpty()}"
    }
```

저는 그동안 `=`을 쓰면 무조건 한 줄로 작성해야 하는 줄 알고, 가독성이 떨어질까 봐 망설였던 것 같아요. 그런데 이렇게 표현하면 `return`을 쓰지 않으면서도 충분히 구조가 명확하고, 코틀린의 간결한 문법도 잘 살릴 수 있겠다는 생각이 들었습니다!

덕분에 표현식 작성에 대해 좀 더 열린 시각을 갖게 되었어요. 앞으로도 습관적으로 `return`이 나올 수 있겠지만 😅 이런 부분 보이면 언제든 편하게 피드백 주세요! 적극적으로 반영해보겠습니다. 감사합니다 💙

leeeyubin commented on src/main/kotlin/com/terning/server/kotlin/ui/api/ExceptionHandler.kt:
```diff
@@ -0,0 +1,70 @@
+import org.springframework.web.context.request.WebRequest
+import org.springframework.web.servlet.mvc.method.annotation.ResponseEntityExceptionHandler
+
+@RestControllerAdvice
+class ExceptionHandler : ResponseEntityExceptionHandler() {
+    override fun handleHttpMessageNotReadable(
```
에러 처리 관련해서도 궁금한 게 많네요.. 가장 궁금한 부분은 현재는 스프링에서 제공하는 기본 에러 타입들을 넣어둔 것 같은데 
저희 프로젝트 도메인에 맞는 에러처리는 따로 안 하고 있는 상황인가요??

jsoonworld replied:
피드백 감사합니다! 😊

도메인에 맞는 에러 처리는 앞으로 꼭 진행해야 할 부분이라고 생각해요!
현재는 각 도메인에서 발생한 예외를 공통된 계층(Global Exception Handler)에서 처리하는 구조로 잡아두었고, 스프링 내부에서 발생하는 에러들은 글로벌에서 일괄적으로 처리하는 느낌이라고 보시면 될 것 같아요!

앞으로 유저 도메인부터 설계를 진행할 예정인데요!
그 과정에서 “도메인에 맞는 예외가 어떤 것이 있을까?“를 고민하면서, 도메인마다 정의하고 처리하는 예외 구조를 함께 정리해보면 좋을 것 같아요.

또, 예외뿐 아니라 각 도메인에서의 성공 응답 처리 방식도 참고해서 일관된 응답 구조를 가져갈 수 있도록 구현할 생각을 갖고 있어요! 🚀

혹시 생각나는 아이디어나 더 좋은 방향이 있다면 편하게 말씀 주세요!

leeeyubin commented on src/main/kotlin/com/terning/server/kotlin/ui/api/ExceptionHandler.kt:
```diff
@@ -0,0 +1,70 @@
+        request: WebRequest
+    ): ResponseEntity<Any> {
+        logger.error("message", ex)
+        val message = when (val exception = ex.cause) {
+            is MismatchedInputException -> "${exception.path.lastOrNull()?.fieldName ?: "UnknownField"}: 널이어서는 안됩니다"
+            is InvalidFormatException -> "${exception.path.lastOrNull()?.fieldName ?: "UnknownField"}: 올바른 형식이어야 합니다"
```
저는 하드코딩은 최대한 지양하려고 하는 편인데 서버 로직을 작성할 때 보통 중요하게 고려하지 않는 편인지도 궁금하네용..

jsoonworld replied:
피드백 감사합니다! 😊 하드코딩에 대한 관점 공유해주셔서 덕분에 저도 한 번 더 생각해보게 되었어요.

저도 지금까지는 하드코딩을 최대한 지양하면서,모두 상수로 전환해서 관리해왔는데요!
최근에는 이런 방식에도 다시 한번 고민이 생겼습니다.

1. 가시성

만약 메시지를 한글로 작성한다고 가정했을 때, 단순히 한 줄짜리 영어 메시지를 상수로 빼는 것보다 하드코딩이 오히려 코드의 의도가 더 잘 보이지 않나? 라는 생각이 들더라고요.
특히 예외 메시지나 한두 번만 쓰이는 안내 문구처럼 치환 이상의 의미가 없을 때는 하드코딩이 더 가독성이 좋다고 느끼기도 했습니다.

2. 재사용성 & 유지보수

반면, 하드코딩을 상수로 관리하는 이유는 결국 재사용성과 유지보수에 있다고 생각해요.
예를 들어 예외 메시지 같은 경우, 테스트에서도 사용하고 실제 서비스 코드에서도 반복된다면,
공통 인터페이스나 상수로 정의해두는 것이 훨씬 효율적이라고 느꼈습니다.

그런데 현재 구조에서는 테스트 코드가 존재하지 않고, 메시지를 사용하는 범위도 서비스 코드 하나뿐이라면, 상수로 분리하기보다는 가시성을 높이는 방식이 더 낫지 않을까? 하는 생각도 듭니다.

혹시 유빈님은 이런 상황에서 어떻게 접근하셨는지, 실제로 하드코딩을 지양하면서 효율을 크게 체감하셨던 경험이 있다면 공유해주시면 감사하겠습니다! 🙏

유빈님의 스타일도 궁금해요!

leeeyubin replied:
장순님 의견을 듣고 생각해봤는데 클라이언트랑 서버랑 비슷한 것 같네요!
단순 에러 메세지만 작성하는 건 저도 하드코딩했던 것 같아요!

그 외에 재사용 되는 부분이 있거나 UI에 나타내야 하는 중요 문자열들은 상수로 관리하곤 했어요.
혹은 1이라는 단순 숫자라도 '화면이 딜레이가 되어야 하는 초 단위'라면 상수로 나타내서 유지보수하기 좋게 했었어요!

현재 이 부분은 따로 재사용을 하지 않기 때문에 가시성을 택하는 방향에 동의합니다!

leeeyubin reviewed (approved):
너무 수고하셨습니다!!
궁금한 점이 많았지만 이 부분은 제가 따로 더 공부해보도록 할게요!!

jsoonworld committed:
refactor(ApiResponse) : unify response format and exclude null fields

jsoonworld committed:
refactor(ExceptionHandler) : standardize error response using unified ApiResponse

- Updated all exception handlers to return ApiResponse with status, message, and nullable result
- Replaced legacy ApiResponse.error(message) with structured version including HTTP status
- Improved consistency of error responses across validation, parsing, and server errors
- Enhanced message formatting for MethodArgumentNotValidException

jsoonworld committed:
refactor(ExceptionHandler) : MethodArgumentNotValidException 메시지 표현식을 간결한 방식으로 리팩터링
